compiler/rustc_trait_selection/src/traits/error_reporting/type_err_ctxt_ext.rs

```diff
@@ -959,34 +961,95 @@ impl<'tcx> TypeErrCtxtExt<'tcx> for TypeErrCtxt<'_, 'tcx> {
+            && let Some(found_kind) = trait_ref.skip_binder().args.type_at(0).to_opt_closure_kind()
+            && let Some(expected_kind) =
+                trait_ref.skip_binder().args.type_at(1).to_opt_closure_kind()
             && !found_kind.extends(expected_kind)
-            && let sig = closure_args.as_closure().sig()
-            && self.can_sub(
-                obligation.param_env,
-                trait_ref,
-                sig.map_bound(|sig| {
-                    ty::TraitRef::new(
-                        self.tcx,
-                        trait_ref.def_id(),
-                        [trait_ref.self_ty().skip_binder(), sig.inputs()[0]],
-                    )
-                }),
-            )
         {
-            let mut err =
-                self.report_closure_error(&obligation, closure_def_id, found_kind, expected_kind);
-            self.note_obligation_cause(&mut err, &obligation);
-            self.point_at_returns_when_relevant(&mut err, &obligation);
-            Some(err.emit())
-        } else {
-            None
+            if let Some((_, Some(parent))) = obligation.cause.code().parent() {
+                trait_ref = parent.to_poly_trait_ref();
+            } else if let &ObligationCauseCode::FunctionArgumentObligation { arg_hir_id, .. } =
+                obligation.cause.code()
+                && let Some(typeck_results) = &self.typeck_results
+                && let ty::Closure(closure_def_id, _) | ty::CoroutineClosure(closure_def_id, _) =
+                    *typeck_results.node_type(arg_hir_id).kind()
+            {
+                let mut err = self.report_closure_error(
+                    &obligation,
+                    closure_def_id,
+                    found_kind,
+                    expected_kind,
+                    "async ",
+                );
+                self.note_obligation_cause(&mut err, &obligation);
+                self.point_at_returns_when_relevant(&mut err, &obligation);
+                return Some(err.emit());
+            }
         }
+
+        let self_ty = trait_ref.self_ty().skip_binder();
+
+        'outer: {
+            if let Some(expected_kind) = self.tcx.fn_trait_kind_from_def_id(trait_ref.def_id()) {
+                let (closure_def_id, found_args, by_ref_captures) = match *self_ty.kind() {
+                    ty::Closure(def_id, args) => {
+                        (def_id, args.as_closure().sig().map_bound(|sig| sig.inputs()[0]), None)
+                    }
+                    ty::CoroutineClosure(def_id, args) => (
+                        def_id,
+                        args.as_coroutine_closure()
+                            .coroutine_closure_sig()
+                            .map_bound(|sig| sig.tupled_inputs_ty),
+                        Some(args.as_coroutine_closure().coroutine_captures_by_ref_ty()),
+                    ),
+                    _ => break 'outer,
```

Comment:
aren't these `break 'outer` just `return None`?

Reply:
or are you planning to add more things below? May be better to just pull out the `'outer` block into a separate function

Reply:
No, I was *planning* to add more things, but now it's totally redundant. Sorry; thanks for catching 😃